\documentclass[a4paper, 12pt]{article}
\usepackage[pdftex]{graphicx}
\usepackage{float}
\usepackage{hyperref}
\usepackage{booktabs}
\usepackage[usenames,dvipsnames]{color}
\usepackage{epstopdf}

\setlength{\voffset}{-0.1in} % set the space between top edge of the paper and the header
\setlength{\textheight}{10in} % set the height of the text body
\setlength{\textwidth}{6.5in} % set the width of the text body

\setlength{\topmargin}{0in}
\setlength{\headheight}{0in}
\setlength{\headsep}{0in}
\setlength{\footskip}{0in}
\setlength{\oddsidemargin}{0in}
\setlength{\evensidemargin}{0in}
\setlength{\parindent}{0in}
\setlength{\parskip}{0.25in}

%\setlength{\topmargin}{0in}
%\setlength{\textheight}{9in}
%\setlength{\oddsidemargin}{0in}
%\setlength{\textwidth}{6.25in}

\hypersetup{
colorlinks=false,
linkcolor={dark-red},
%citecolor={dark-blue},
%urlcolor={medium-blue},
pdfborder={0 0 0},
}

\begin{document}
\begin{figure}
\centering
\href{http://www.sequencing.uio.no/}{\includegraphics[scale=0.35]{__TemplateDir__/NSC_logo_original_CMYK.pdf}}
\end{figure}

% Title, Author and Explanation
\title{\bf \Huge Quality Control Report}
%\author{(\href{http://www.sequencing.uio.no/}{http://www.sequencing.uio.no/})}
\maketitle
This report aims to give a brief overview of the quality of the sequence produced from your sample and should be self-explanatory. However, you may want to refer to our website (\href{http://www.sequencing.uio.no/services/data-delivery}{http://www.sequencing.uio.no/services/data-delivery}) for further details on sequence and quality control files.
\\
\\
The graphs in the report were generated using fastqc (\href{http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/} {http://www.bioinformatics.bbsrc.ac.uk/}). Fastqc generates many different reports and only a subset of them are included here. If you wish to see the full set, please download the software and run it on your sequence files.

% First section, Run Summary (table)
\section{Run summary}
\begin{center}
\begin{table}[H]
\centering
\begin{tabular}{lclcl}
\toprule
{\bf Measure} &{\bf Value}\\
\midrule
{\bf Run name} & __RunName__\\
{\bf Run mode} & __RunMode__ \\
{\bf Chemistry Version} & __ChemistryVersion__ \\
{\bf __ControlSoftwareName__ Version } & __ControlSoftwareVersion__ \\
{\bf RTA Version} & __RTAVersion__ \\
{\bf bcl2fastq Version} & __bcl2fastqVersion__\\
{\bf Sample name} & __SampleName__ \\
{\bf Read} & __ReadNum__ \\
%{\bf Average number of clusters per tile} & @@ClusterN@@\\
{\bf Total number of sequences} & __TotalN__\\
\bottomrule
\end{tabular}
\end{table}
\end{center}
%need to change the table

\newpage
\section {Per Base Sequence Quality}
The graph shows an overview of the range of quality values across all bases at each position in the FastQ file. The central red line shows the median of the quality values and the blue line shows the mean of the quality values.
\begin{figure}[H]
\centering
\includegraphics[width=6.5in,height=3in]{__Folder__/Images/per_base_quality.png}
\end{figure}
%need to introduce the figure

\section{Per Sequence Quality Scores}
The graph is generated by computing the average quality of a read (by averaging across read positions) and then plotting the distribution of this average quality. It thus enables you to tell whether low quality bases are located in a subset of the reads or distributed across all reads.

\begin{figure}[H]
\centering
\includegraphics[scale=0.35]{__Folder__/Images/per_sequence_quality.png}
\end{figure}
%need to introduce the figure

\newpage
\section{Per Base Sequence Content}
The graph plots the proportion of each base type in each read position. 
\begin{figure}[H]
\centering
\includegraphics[width=6.5in,height=3in]{__Folder__/Images/per_base_sequence_content.png}
\end{figure}
%need to introduce the figure

\section{Per Base N Content}
The graph shows the percentage of base calls at each position for which an N was called. If a sequencer is unable to make a base call with sufficient confidence then it will normally call an N rather than A, T, G or C. 
\begin{figure}[H]
\centering
\includegraphics[width=6.5in,height=3in]{__Folder__/Images/per_base_n_content.png}
\end{figure}

\newpage
\section{Sequence Duplication levels}
The graph shows the number of sequences with different degrees of duplication relative to the number of unique sequences (which is set to 100\%). In a diverse library, most sequences will occur only once in the final set and the graph will show a peak in the unique category. However, some sequences may be present in more than one copy (for example, as the result of PCR amplification), in which the graph may show high numbers of sequences in the other categories (2 copies, 3 copies, etc). The last category is for 10 copies or more.
 
\begin{figure}[H]
\centering
\includegraphics[scale=0.35]{__Folder__/Images/duplication_levels.png}
\end{figure}

\end{document}
